sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -656,6 +659,73 @@ trait ParquetPartitionReaderBase extends Logging with Arm with ScanWithMetrics
     }
   }
 
+  /**
+   * Convert cudf unsigned integer to wider signed integer that parquet expects
+   * After spark 3.2.0, parquet read uint8 as int16, uint16 as int32, uint32 as int64
+   *
+   * @param inputTable
+   * @return
+   */
+  private def convertToUnsigned(inputTable: Table) = {
+    var typeCastingNeeded = needUnsignedCast(inputTable, readDataSchema)
+    if (typeCastingNeeded) {
+      val newColumns = new Array[ColumnVector](readDataSchema.length)
+      try {
+        withResource(inputTable) { table =>
+          (0 until readDataSchema.length).foreach(i => {
+            val readField = readDataSchema(i)
+            val origCol = table.getColumn(i)
+            val col: ColumnVector =
+              ColumnCastUtil.ifTrueThenDeepConvertTypeAtoTypeB(origCol, readField.dataType,
+                (dt, cv) => needUnsignedToSignedCastFunction(cv, dt),
+                (dt, cv) => cv.castTo(
+                  DType.create(GpuColumnVector.getNonNestedRapidsType(dt).getTypeId)))
+            newColumns(i) = col
+          })
+        }
+        new Table(newColumns: _*)
+      } finally {
+        newColumns.safeClose()
+      }
+    } else {
+      inputTable
+    }
+  }
+
+ def needUnsignedCast(table: Table, schema: StructType): Boolean = {
+    var b = false
+    for (i <- 0 until table.getNumberOfColumns) {
+      if (needUnsignedToSignedCast(table.getColumn(i), schema.fields(i).dataType)) {
+        b =true
+      }
+    }
+    b
+  }
+
+  def needUnsignedToSignedCast(cv: ColumnView, dataType: DataType): Boolean = {
+    dataType match {
+      case a: ArrayType =>
+        val child = cv.getChildColumnView(0)
+        return needUnsignedToSignedCast(child, a.elementType)
+      case s: StructType =>
+        for( i <- 0 until cv.getNumChildren) {
+          val child = cv.getChildColumnView(i)
+          if (needUnsignedToSignedCast(child, s(i).dataType)) {
+            return true
+          }
+        }
+      case _ =>
+        return needUnsignedToSignedCastFunction(cv, dataType)
+    }
+    return false
+  }
+
+  def needUnsignedToSignedCastFunction(cv: ColumnView, dataType: DataType): Boolean = {
```

Comment:
Personally, it is not neccessary to be an independent method. It would be good to move it into `needUnsignedToSignedCast` .

Reply:
this is need in the "(dt, cv) => " function